File: backend/app/services/clinical_service.py
from __future__ import annotations

import uuid

from sqlalchemy import select
from sqlalchemy.orm import Session

from app.data import CHIEF_COMPLAINT_QUESTION, COMPLAINTS, is_complaint_id
from app.models import (
    Allergy,
    ClinicalAlert,
    ClinicalHistory,
    Interview,
    Medication,
    Patient,
    TimelineEvent,
    VitalObservation,
)
from app.schemas.clinical import ClinicalHistoryPatch
from app.services import ai_service


def get_opening_question() -> dict:
    return CHIEF_COMPLAINT_QUESTION


def get_follow_up_questions(complaint_id: str) -> list[dict]:
    if not is_complaint_id(complaint_id):
        return []
    return COMPLAINTS[complaint_id]["followUps"]


def _history_to_out(history: ClinicalHistory) -> dict:
    return {
        "patientId": history.patient_id,
        "chiefComplaint": history.chief_complaint,
        "historyOfPresentIllness": history.history_of_present_illness,
        "pastMedicalHistory": history.past_medical_history,
        "pastSurgicalHistory": history.past_surgical_history,
        "medications": [
            {"id": m.id, "name": m.name, "dose": m.dose, "frequency": m.frequency} for m in history.medications
        ],
        "allergies": [
            {"id": a.id, "substance": a.substance, "reaction": a.reaction, "severity": a.severity}
            for a in history.allergies
        ],
        "familyHistory": history.family_history,
        "personalHistory": history.personal_history,
        "reviewOfSystems": history.review_of_systems,
        "investigationsSummary": history.investigations_summary,
        "keySymptoms": history.key_symptoms or [],
        "riskIndicators": history.risk_indicators or [],
        "suggestedQuestions": history.suggested_questions or [],
        # `or ""` covers rows written before the AI-summary migration
        # (b28d5f1a9c6e), which have NULL here rather than "" — the column
        # is nullable (MySQL disallows a TEXT column DEFAULT), so this can't
        # rely on a DB-level default.
        "clinicalSummary": history.clinical_summary or "",
        "aiGenerated": history.ai_generated,
        "confirmedByClinician": history.confirmed_by_clinician,
        "updatedAt": history.updated_at,
    }


def get_history(db: Session, patient_id: str) -> dict | None:
    history = db.get(ClinicalHistory, patient_id)
    return _history_to_out(history) if history else None


def save_history(db: Session, patient_id: str, patch: ClinicalHistoryPatch) -> dict:
    history = db.get(ClinicalHistory, patient_id)
    if not history:
        history = ClinicalHistory(patient_id=patient_id)
        db.add(history)

    data = patch.model_dump(exclude_unset=True, exclude={"medications", "allergies"})
    for field, value in data.items():
        setattr(history, field, value)

    if patch.medications is not None:
        history.medications.clear()
        for med in patch.medications:
            history.medications.append(
                Medication(
                    id=med.id or f"med-{uuid.uuid4().hex[:8]}",
                    patient_id=patient_id,
                    name=med.name,
                    dose=med.dose,
                    frequency=med.frequency,
                )
            )

    if patch.allergies is not None:
        history.allergies.clear()
        for allergy in patch.allergies:
            history.allergies.append(
                Allergy(
                    id=allergy.id or f"all-{uuid.uuid4().hex[:8]}",
                    patient_id=patient_id,
                    substance=allergy.substance,
                    reaction=allergy.reaction,
                    severity=allergy.severity,
                )
            )

    db.commit()
    db.refresh(history)
    return _history_to_out(history)


def generate_ai_summary(db: Session, patient_id: str) -> dict | None:
    """Assembles what's already on file for this patient — complaint,
    existing history, structured interview answers, and the deterministic
    triage flags — and asks `ai_service` for a structured clinical summary.

    Read-only: this never writes to the database by itself. The doctor
    reviews the draft on `/doctor/summary` and only Save/Confirm & Save
    (the existing PATCH endpoint) persists anything.
    """
    patient = db.get(Patient, patient_id)
    if not patient:
        return None

    history = db.get(ClinicalHistory, patient_id)
    interview = db.scalars(
        select(Interview).where(Interview.patient_id == patient_id).order_by(Interview.created_at.desc())
    ).first()

    raw_answers = []
    complaint_id = interview.complaint_id if interview else None
    if interview:
        raw_answers = [
            {"questionId": a.question_id, "optionIds": a.option_ids, "transcript": a.transcript}
            for a in interview.answers
        ]
    qa_pairs = ai_service.resolve_answer_labels(complaint_id, raw_answers)

    result = ai_service.generate_clinical_summary(
        complaint=patient.complaint,
        complaint_id=complaint_id,
        history_of_present_illness=history.history_of_present_illness if history else "",
        qa_pairs=qa_pairs,
        risk_flags=patient.flags or [],
    )
    return result


def get_timeline(db: Session, patient_id: str) -> list[dict]:
    events = db.scalars(
        select(TimelineEvent).where(TimelineEvent.patient_id == patient_id).order_by(TimelineEvent.sort_key.desc())
    ).all()
    return [
        {
            "id": e.id,
            "patientId": e.patient_id,
            "date": e.date,
            "sortKey": e.sort_key,
            "type": e.type,
            "icon": e.icon,
            "color": e.color,
            "title": e.title,
            "summary": e.summary,
            "detail": e.detail,
            "abnormal": e.abnormal,
            "source": e.source,
        }
        for e in events
    ]


def get_alerts(db: Session, patient_id: str) -> list[dict]:
    alerts = db.scalars(select(ClinicalAlert).where(ClinicalAlert.patient_id == patient_id)).all()
    return [
        {
            "id": a.id,
            "patientId": a.patient_id,
            "category": a.category,
            "severity": a.severity,
            "title": a.title,
            "value": a.value,
            "referenceRange": a.reference_range,
            "status": a.status,
            "date": a.date,
            "note": a.note,
        }
        for a in alerts
    ]


def get_vitals(db: Session, patient_id: str) -> list[dict]:
    vitals = db.scalars(select(VitalObservation).where(VitalObservation.patient_id == patient_id)).all()
    return [{"id": v.id, "label": v.label, "value": v.value, "unit": v.unit, "normal": v.normal} for v in vitals]

File: src/services/clinicalService.ts
import {
  CHIEF_COMPLAINT_QUESTION,
  COMPLAINTS,
  MOCK_ALERTS,
  MOCK_HISTORIES,
  MOCK_TIMELINE,
  MOCK_VITALS,
} from "@/data"
import type {
  AdaptiveQuestion,
  AdaptiveQuestionTurn,
  AiClinicalSummary,
  Answer,
  ClinicalAlert,
  ClinicalHistory,
  ComplaintId,
  Question,
  RedFlagAssessment,
  TimelineEvent,
  VitalObservation,
} from "@/types"
import { evaluateRedFlags } from "@/utils"
import { API_CONFIG, delay, request } from "./apiClient"

/** Local overrides so clinician edits survive navigation during a session. */
const historyEdits = new Map<string, ClinicalHistory>()

export const clinicalService = {
  /** The opening question. Constant today; backend-driven later. */
  async getOpeningQuestion(): Promise<Question> {
    if (!API_CONFIG.useMock) return request<Question>("/interview/opening")
    await delay(60)
    return CHIEF_COMPLAINT_QUESTION
  },

  /**
   * Adaptive follow-ups for a chief complaint.
   *
   * The demo reads a static bank. The FastAPI/LLM endpoint will return the same
   * `Question[]` shape, conditioned on the answers given so far — which is why
   * `previousAnswers` is already part of the signature.
   */
  async getFollowUpQuestions(
    complaintId: ComplaintId,
    previousAnswers: Answer[] = [],
  ): Promise<Question[]> {
    if (!API_CONFIG.useMock) {
      return request<Question[]>("/interview/follow-ups", {
        method: "POST",
        body: { complaintId, previousAnswers },
      })
    }
    await delay(220)
    return COMPLAINTS[complaintId].followUps
  },

  /**
   * Demo triage evaluation. Runs locally so the kiosk keeps working offline;
   * the backend will own this once clinical governance is in place.
   */
  async assessRedFlags(
    complaintId: ComplaintId | null,
    answers: Answer[],
  ): Promise<RedFlagAssessment> {
    if (!API_CONFIG.useMock) {
      return request<RedFlagAssessment>("/triage/assess", {
        method: "POST",
        body: { complaintId, answers },
      })
    }
    await delay(150)
    return evaluateRedFlags(complaintId, answers)
  },

  async getHistory(patientId: string): Promise<ClinicalHistory | undefined> {
    if (!API_CONFIG.useMock) {
      return request<ClinicalHistory>(`/patients/${patientId}/history`)
    }
    await delay(120)
    return historyEdits.get(patientId) ?? MOCK_HISTORIES[patientId]
  },

  async saveHistory(
    patientId: string,
    patch: Partial<ClinicalHistory>,
  ): Promise<ClinicalHistory> {
    // THE SAVE BUG: this used to build `base`/`updated` — which only exist to
    // support the mock branch below — before checking `useMock` at all. Since
    // `historyEdits` starts empty and a real patient id is never a key in
    // `MOCK_HISTORIES`, that lookup's fallback threw on every single real-mode
    // save, regardless of whether the PATCH request that follows would have
    // succeeded. The FastAPI endpoint and MySQL side were always correct —
    // this function just never reached the `request()` call. Checking the
    // mode first, and only doing the mock-local computation inside the mock
    // branch, is the fix.
    if (!API_CONFIG.useMock) {
      return request<ClinicalHistory>(`/patients/${patientId}/history`, {
        method: "PATCH",
        body: patch,
      })
    }

    const base =
      historyEdits.get(patientId) ??
      MOCK_HISTORIES[patientId] ??
      (() => {
        throw new Error(`No history for ${patientId}`)
      })()

    const updated: ClinicalHistory = {
      ...base,
      ...patch,
      patientId,
      updatedAt: new Date().toISOString(),
    }

    await delay(220)
    historyEdits.set(patientId, updated)
    return updated
  },

  /** Marks the AI draft as reviewed by a clinician. */
  async confirmHistory(patientId: string): Promise<ClinicalHistory> {
    return clinicalService.saveHistory(patientId, { confirmedByClinician: true })
  },

  async getTimeline(patientId: string): Promise<TimelineEvent[]> {
    if (!API_CONFIG.useMock) {
      return request<TimelineEvent[]>(`/patients/${patientId}/timeline`)
    }
    await delay(140)
    return [...(MOCK_TIMELINE[patientId] ?? [])].sort((a, b) =>
      b.sortKey.localeCompare(a.sortKey),
    )
  },

  async getAlerts(patientId: string): Promise<ClinicalAlert[]> {
    if (!API_CONFIG.useMock) {
      return request<ClinicalAlert[]>(`/patients/${patientId}/alerts`)
    }
    await delay(120)
    return MOCK_ALERTS[patientId] ?? []
  },

  async getVitals(patientId: string): Promise<VitalObservation[]> {
    if (!API_CONFIG.useMock) {
      return request<VitalObservation[]>(`/patients/${patientId}/vitals`)
    }
    await delay(90)
    return MOCK_VITALS[patientId] ?? []
  },

  /**
   * Generates an AI-assisted clinical summary draft (Gemini on the backend;
   * see backend/app/services/ai_service.py). Read-only — this never saves
   * anything by itself. The doctor reviews the returned draft on
   * `/doctor/summary`, edits it, and only Save/Confirm & Save (`saveHistory`
   * above) persists it, through the one existing PATCH endpoint.
   *
   * `source` is always "ai" or "fallback" — never fake "AI-generated" status
   * when Gemini didn't actually run.
   */
  async generateAiSummary(patientId: string): Promise<AiClinicalSummary> {
    if (!API_CONFIG.useMock) {
      return request<AiClinicalSummary>(`/patients/${patientId}/ai-summary`, {
        method: "POST",
      })
    }
    await delay(300)
    const history = historyEdits.get(patientId) ?? MOCK_HISTORIES[patientId]
    if (!history) {
      return {
        chiefComplaint: "",
        historyPresentIllness: "",
        keySymptoms: [],
        riskIndicators: [],
        suggestedQuestions: [],
        clinicalSummary: "",
        source: "fallback",
      }
    }
    return {
      chiefComplaint: history.chiefComplaint,
      historyPresentIllness: history.historyOfPresentIllness,
      keySymptoms: [],
      riskIndicators: [],
      suggestedQuestions: [],
      clinicalSummary: `Patient reports: ${history.chiefComplaint}. ${history.historyOfPresentIllness}`,
      source: "fallback",
    }
  },

  /**
   * Next adaptive follow-up question for the kiosk interview (Gemini). Never
   * used to diagnose or determine triage — the deterministic red-flag rules
   * in `assessRedFlags` above are unaffected by this and remain the sole
   * source of P1/P2/P3. On any failure (including mock mode, which has no
   * LLM to call), `source` is "fallback" and `question` is null — callers
   * fall back to `getFollowUpQuestions`'s predefined bank in that case.
   */
  async getAdaptiveQuestion(
    complaint: string,
    language: string,
    history: AdaptiveQuestionTurn[],
  ): Promise<AdaptiveQuestion> {
    if (!API_CONFIG.useMock) {
      return request<AdaptiveQuestion>("/ai/adaptive-question", {
        method: "POST",
        body: { complaint, language, history },
      })
    }
    await delay(200)
    return { question: null, questionNumber: history.length + 1, isFinal: true, source: "fallback" }
  },

  /** Test/demo helper — drops in-session clinician edits. */
  resetEdits(): void {
    historyEdits.clear()
  },
}
